packages/pluggableWidgets/barcode-scanner-web/src/components/BarcodeScanner.tsx

```diff
@@ -0,0 +1,119 @@
+import { createElement, ReactElement, useCallback, useEffect, useRef, useState } from "react";
+import classNames from "classnames";
+import { Alert } from "@mendix/piw-utils-internal";
+import { useCodeScanner, CodeScannerHookError } from "../hooks/useCodeScanner";
+import { browserSupportsCameraAccess, useMediaStream, MediaStreamHookError } from "../hooks/useMediaStream";
+
+import "../ui/BarcodeScanner.scss";
+
+export interface BarcodeScannerProps {
+    onClose?: () => void;
+    onDetect?: (data: string) => void;
+    showMask: boolean;
+}
+
+function getErrorMessage(errorEnum: MediaStreamHookError | CodeScannerHookError | null): string | null {
+    switch (errorEnum) {
+        case "ERROR_NOT_FOUND":
+            return "Error in barcode scanner: no camera media devices were found.";
+        case "ERROR_MEDIA_STREAM":
+            return "Error in barcode scanner: an unexpected error occurred while retrieving the camera media stream.";
+        case "ERROR_CODE_SCANNER":
+            return "Error in barcode scanner: an unexpected error occurred while detecting a barcode in the camera media stream.";
+        case "ERROR_NOT_ALLOWED":
+        default:
+            return null;
+    }
+}
+
+export function BarcodeScanner({ onClose, onDetect, showMask }: BarcodeScannerProps): ReactElement | null {
+    const [showScannerOverlay, setShowScannerOverlay] = useState<boolean>(true);
+    const [videoElement, setVideoElement] = useState<HTMLVideoElement | null>(null);
+    const hasDetectedOnce = useRef<boolean>(false);
+    const { streamObject, cleanupStreamObject, error: errorMediaStream } = useMediaStream();
+    const { codeResult, error: errorCodeScanner } = useCodeScanner(streamObject, videoElement);
+    const supportsCameraAccess = browserSupportsCameraAccess();
+
+    const updateVideoElement = useCallback(
```

Comment:
Can we use `useRef` instead?

Reply:
Isa and I came across an issue with `useRef` in jest env not triggering certain lifecycle methods. In this case we're also dealing with a ref to a html element that can easily affect UI, so I think putting it in state should be fine. But if you want, I can take a look whether using ref is also fine here

Reply:
What i would like to prevent is the re-render being triggered by the `setVideoElement` as this one is inside the state.

Reply:
I also don't think make sense to store an HTML element in the state instead of a ref.. which is the feature you need.

Reply:
The [React documentation on `useRef`](https://reactjs.org/docs/hooks-reference.html#useref) says the following:

> Keep in mind that useRef doesn’t notify you when its content changes. Mutating the .current property doesn’t cause a re-render. If you want to run some code when React attaches or detaches a ref to a DOM node, you may want to use a callback ref instead.

It's just a coincidence that the ref is set in the browser, when we just use refs. That's what we found out when it was failing in Jest tests with Enzyme.

`setVideoElement` is ensured to stay the same throughout re-renders. We just add it to the dependency list for completeness and the no exhaustive deps rule.

> React guarantees that setState function identity is stable and won’t change on re-renders. This is why it’s safe to omit from the useEffect or useCallback dependency list.

Reply:
`useState` here is being used as `useRef`. We are ust not using `useRef` because of Jest, which should be fixed instead of do workaround to prevent errors while running unit tests..

Reply:
Changed it to `useRef`. Tests still pass and couldn't find any regression when going through it locally and manually